File: burst-balloons.java
class Solution {
    //TC: O(n^3)
    //SC: O(n*n)
    public int maxCoins(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        int n = nums.length;
        int [][] dp = new int[n][n];
        // Create length of subProblem
        for(int g = 0; g < n; g++){
            for(int i = 0, j = g; j < n; j++, i++){
                int max = Integer.MIN_VALUE;
                for(int k = i; k <=j ; k++)
                {
                    int left = k == i ? 0: dp[i][k-1];
                    int right = k == j ? 0 : dp[k+1][j];
                    int val = (i == 0 ? 1: nums[i-1]) * nums[k] * (j == n-1 ? 1 :nums[j+1]);
                    
                    int total = left + right + val;
                    max = Math.max(max,total);
                }
                dp[i][j] = max;
            }
        }


        return dp[0][n-1];
        

    }
}

File: BurstBalloons.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace DynamicProgramming
{
    public class BurstBalloons
    {
        /*
         * T.C: O(n2)
         * S.C: O(n2)
         */
        public int MaxCoins(int[] nums)
        {
            if (nums == null || nums.Length == 0) return 0;

            int n = nums.Length;
            int[,] dp = new int[n, n];

            for (int len = 1; len <= n; len++)
            {
                //starting index
                for (int i = 0; i <= n - len; i++)
                {
                    int j = i + len - 1;
                    for (int k = i; k <= j; k++)
                    {
                        int before = 0;
                        if (k != i)
                        {
                            before = dp[i, k - 1];
                        }
                        int after = 0;
                        if (k != j)
                        {
                            after = dp[k + 1, j];
                        }

                        int left = 1;
                        if (i != 0)
                        {
                            left = nums[i - 1];
                        }
                        int right = 1;
                        if (j != n - 1)
                        {
                            right = nums[j + 1];
                        }
                        dp[i, j] = Math.Max(dp[i, j], before + left * nums[k] * right + after);
                    }
                }
            }

            return dp[0, n - 1];


        }
    }
}
